src/containers/BubbleEditor/BubbleEditor.js

```diff
@@ -41,13 +44,95 @@ class BubbleEditor extends React.Component {
         _timePoints.add(bubble[RANGE.END_TIME]);
         return _timePoints;
       }, new Set())
+    ).sort((p1, p2) => p1 - p2);
+
+  clearTextSelection = () => {
+    if (window.getSelection) {
+      if (window.getSelection().empty) {
+        // Chrome
+        window.getSelection().empty();
+      } else if (window.getSelection().removeAllRanges) {
+        // Firefox
+        window.getSelection().removeAllRanges();
+      }
+    } else if (document.selection) {
+      // IE?
+      document.selection.empty();
+    }
+  };
+
+  dragStart = ev => {
+    if (
+      ev.target === ev.target.parentNode.firstChild ||
+      ev.target === ev.target.parentNode.lastChild
+    ) {
+      return;
+    }
+    const selectedPoint = Array.prototype.indexOf.call(
+      ev.target.parentNode.childNodes,
+      ev.target
     );
 
+    document.body.addEventListener('mousemove', this.dragMove);
+    document.body.addEventListener('mouseup', this.dragEnd);
+    this.setState({
+      selectedPoint: selectedPoint,
+      startX: ev.pageX,
+      deltaX: 0,
+    });
+  };
+
+  dragMove = ev => {
+    if (this.state.selectedPoint < 0) {
+      return;
+    }
+    // in order to smooth drag
+    this.clearTextSelection();
+    this.setState({
+      deltaX: ev.clientX - this.state.startX,
+    });
+  };
+
+  dragEnd = ev => {
+    if (this.state.selectedPoint !== -1) {
+      document.body.removeEventListener('mousemove', this.dragMove);
+      document.body.removeEventListener('mouseup', this.dragEnd);
+      const originalX = this.getTimePoints()[this.state.selectedPoint];
+      const dX =
+        ((ev.pageX - this.state.startX) / this.state.dimensions.width) *
+        this.props.runTime;
+
+      this.props.movePoint(originalX + dX, originalX);
+    }
+    this.setState({
+      selectedPoint: -1,
+    });
+  };
+
   render() {
     const _points = this.props.points;
     const { runTime, currentTime, zoom, onUpdateTime, splitRange } = this.props;
+    const { dimensions, selectedPoint, deltaX } = this.state;
     const timePoints = this.getTimePoints();
 
+    let selectedPointValue = 0;
+    let substituteValue = 0;
+    if (selectedPoint !== -1) {
+      selectedPointValue = timePoints[selectedPoint];
+      timePoints[selectedPoint] += Math.max(
+        (deltaX / dimensions.width) * runTime
+      );
+      timePoints[selectedPoint] = Math.max(
+        timePoints[selectedPoint],
+        timePoints[selectedPoint - 1]
+      );
+      timePoints[selectedPoint] = Math.min(
+        timePoints[selectedPoint],
+        timePoints[selectedPoint + 1]
+      );
+      substituteValue = timePoints[selectedPoint];
+    }
```

Comment:
Should this be part of the `this.getTimePoints()` or maybe split into its own function. `substituteValue` and  `selectedPointValue` are local to that scope, and I can't see where they are used below.

Reply:
Because that's the part I'm working on... I pushed it from my home computer, and continue working on it at the office.

Reply:
Next round, we need to demo the existing... I create a bug report ticket for this